`lg_exp_stringify()` fixes

ampli opened the pull request:
Bug fix:
- `print_expression_tag_start()`: Fix printing extra `[` when not printing tags
- 
API change:
- `lg_exp_stringify()`: Return a new memory
- `lg_exp_operand_*()`: Use const for parameter and result

The API changes are according to my proposals in issue #1222. I didn't implement a special `free()`.

`clinkgrammar.lg_exp_stringify()` now leaks memory in Python. This is to be fixed in the next PR (needs a rebase on `master` after applying this PR).

Force-pushed:
-  Change `lg_exp_stringify()` to return `char *` instead of `const char *`, as usual for functions that don't own the resulting memory.
- Change  `exp_stringify()` (the function for internal use, that uses TLS memory for its result) to call an internal function that `lg_exp_stringify()` calls (instead of calling `lg_exp_stringify()`, which has an external linkage).

linas commented:
I don;'t think you need to rebase for a second PR.  That is, the following works fine:
```
git checkout master
git checkout -b branch-a
hack
git commit
git push origin branch-a
# go to github and create pull req for branch-a
git checkout -b branch-more   # This is a continuation of the work on branch-a, e.g. the python work.
hack
git commit
git push origin branch-b
# goto github and create pull req for branch-b. This will look "just fine".
# oops found an error in branch-a
git checkout branch-a
hack
git commit
git push origin branch-a # this will auto-update the pull req. No force-push required.
# the above won;'t change the pull-req for branch-b, but everything will be fine
# after a and b are merged. 
```
You can bounce around between these branches, and more, adding more commits all the time, and it will all work fine, without any need at all for any rebases or any force-pushes,  And if, for some reason, you need to have the latest fix from `branch-a` before you can commit something to branch be, well, that's easy, too:
```
git checkout branch-b
git merge branch-a
hack
git commit
git push
``` 
Still no rebase or force-push required, and everything will look just fine in the github pull reqs; they won't get garbled.

ampli commented:
> I don't think you need to rebase for a second PR
> [...]
> Still no rebase or force-push required

Some comments:

When I fix a code that is not yet merged to `master`, I just continue my regular workflow of removing the bad code altogether by rewriting the commits. The forced-push was needed because I didn't add a commit but changed an existing one. But in that case, I forgot to release beforehand... (see the `fixup!`).

> git merge branch-a

This is similar to a rebase, just leaving a merge record. 


Regarding working on feature branches, my workflow is totally different:

I use worktrees (now using `git worktree` but previous to the worktree integration into `git` I used an external script for that). This means these branches are always checked out (I rarely use `git checkout` for branches). I currently have over **100**(!) such branches, which I consider as "active development". I keep the inactive branches (several hundred) using `git tag -a` for "future reference").

The worktree branches get out of date every time I send a PR... After a PR is applied, I return to the branches from which I intend to prepare the next PRs and rebase them so I can continue to work on an up-to-date code. I could rebase in advance but I would like to be sure I do it with an actual repository code.

There is an additional workflow I regularly use (almost for each PR):

I break branches into several PRs and/or cherry-pick commits for the candidate PR. Then I rebase the original branches so I don't send commits that I already sent (GitHub is not clever when this happens). For example, now I have a branch named `exp-resolve` that I created for adding the `lg_exp_resolve()` API call. But I wanted to make a test for it and to that end, I needed to bind an API call that returns an expression. So I decided to bind `dictionary_lookup_list()`. But I realize I can just bind all the rest of the API at the same time. And then I needed to test the API calls that I need for testing `lg_exp_resolve()`... and meanwhile I found and fixed bugs in the code that I touched.

So I finally got a branch that implements several different things, among them binding the rest of the API calls and implementing a new API `lg_exp_resolve()` with its tests. I already extracted out of it commits that implement fixes/enhancements and sent them separately. Now I intend to break it (by cherry-picks) into a pure `lg_exp_resolve()` implementation and a branch to add the rest of the API.

**Of course, I would appreciate much any idea how to manage working on >100 branches and keep them up to date with less frequent rebasing (or merging).**

(BTW, in old branches that I try to revive ( rebasing becomes too complicated. For example, I have a big difficulty in reviving my expression-simplification branch. I got a buggy result, and it is not easy to debug.)

linas commented:
OK, well, you've got a sophisticated system going, then. 

I have lots of old, stale branches, but effectively all of them are abandoned, and will probably never be revived. Whatever idea I was pursuing in them, it became clear it would not work. Thus, I don't ever need to keep them working and up-to-date.

I explained my workflow mostly because you said that you won't push your next request until the last one is merged. For me, that is never a problem: I can create many pull reqs, each depending on the earlier ones, and they all work out just fine. I think I've done a rebase once in my life. Merges more often, but the need is infrequent.
